File: Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/CSharpierProcessServer.cs
using System;
using System.CodeDom;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.NetworkInformation;
using System.Text;
using System.Threading.Tasks;
using CSharpier.VisualStudio;
using Newtonsoft.Json;

public class CSharpierProcessServer : ICSharpierProcess2, IDisposable
{
    private readonly string csharpierPath;
    private readonly Logger logger;
    private int port;
    private Process? process;
    public bool ProcessFailedToStart { get; }

    public string Version { get; }

    public CSharpierProcessServer(string csharpierPath, string version, Logger logger)
    {
        this.logger = logger;
        this.csharpierPath = csharpierPath;
        this.Version = version;

        if (!this.StartProcess())
        {
            this.ProcessFailedToStart = true;
            return;
        }

        this.logger.Debug("Warm CSharpier with initial format");
        // warm by formatting a file twice, the 3rd time is when it gets really fast
        this.FormatFile("public class ClassName { }", "/Temp/Test.cs");
        this.FormatFile("public class ClassName { }", "/Temp/Test.cs");
    }

    private bool StartProcess()
    {
        try
        {
            const string arguments = "--server";

            this.logger.Debug("Running " + this.csharpierPath + " --server");

            var processStartInfo = new ProcessStartInfo(this.csharpierPath, arguments)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                Environment = { ["DOTNET_NOLOGO"] = "1" },
            };
            this.process = Process.Start(processStartInfo);

            var output = string.Empty;

            var task = Task.Run(() =>
            {
                output = this.process!.StandardOutput.ReadLine();
            });

            if (!task.Wait(TimeSpan.FromSeconds(2)))
            {
                this.process!.Kill();
                this.logger.Warn(
                    "Spawning the csharpier server timed out."
                        + Environment.NewLine
                        + this.process!.StandardError.ReadToEnd()
                );
                return false;
            }

            if (this.process!.HasExited)
            {
                this.logger.Warn(
                    "Spawning the csharpier server failed because it exited. "
                        + this.process!.StandardError.ReadToEnd()
                );
                return false;
            }

            var portString = output.Replace("Started on ", "");
            this.port = int.Parse(portString);

            this.logger.Debug("Connecting via port " + portString);
            return true;
        }
        catch (Exception e)
        {
            this.logger.Warn("Failed to spawn the needed csharpier server." + e);
            return false;
        }
    }

    public string FormatFile(string content, string filePath)
    {
        var parameter = new FormatFileParameter { fileName = filePath, fileContents = content };

        var result = this.formatFile(parameter);
        return result?.formattedFile ?? string.Empty;
    }

    public FormatFileResult? formatFile(FormatFileParameter parameter)
    {
        if (this.ProcessFailedToStart)
        {
            this.logger.Warn("CSharpier process failed to start. Formatting cannot occur.");
            return null;
        }

        var url = "http://127.0.0.1:" + this.port + "/format";

        try
        {
            var request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = "POST";
            request.ContentType = "application/json; charset=utf-8";

            using (var streamWriter = new StreamWriter(request.GetRequestStream()))
            {
                streamWriter.Write(JsonConvert.SerializeObject(parameter));
            }

            var response = (HttpWebResponse)request.GetResponse();

            if (response.StatusCode != HttpStatusCode.OK)
            {
                this.logger.Warn(
                    "Csharpier server returned non-200 status code of " + response.StatusCode
                );
                response.Close();
                return null;
            }

            using (var streamReader = new StreamReader(response.GetResponseStream(), Encoding.UTF8))
            {
                var result = JsonConvert.DeserializeObject<FormatFileResult>(
                    streamReader.ReadToEnd()
                );
                return result;
            }
        }
        catch (Exception e)
        {
            this.logger.Warn("Failed posting to the csharpier server. " + e);
        }

        return null;
    }

    public void Dispose()
    {
        this.process?.Kill();
    }
}

File: Src/CSharpier.Rider/src/main/java/com/intellij/csharpier/CSharpierProcessServer.java
package com.intellij.csharpier;

import com.google.gson.Gson;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class CSharpierProcessServer implements ICSharpierProcess2, Disposable {

    private final Gson gson = new Gson();
    private final String csharpierPath;
    private final DotNetProvider dotNetProvider;
    private final String version;
    private Logger logger = CSharpierLogger.getInstance();
    private int port;
    private Process process = null;
    private boolean processFailedToStart;

    public CSharpierProcessServer(String csharpierPath, String version, Project project) {
        this.csharpierPath = csharpierPath;
        this.dotNetProvider = DotNetProvider.getInstance(project);
        this.version = version;

        this.startProcess();

        this.logger.debug("Warm CSharpier with initial format");
        // warm by formatting a file twice, the 3rd time is when it gets really fast
        this.formatFile("public class ClassName { }", "/Temp/Test.cs");
        this.formatFile("public class ClassName { }", "/Temp/Test.cs");
    }

    private void startProcess() {
        var newCommandsVersion = "1.0.0-alpha1";
        var argument = Semver.gte(this.version, newCommandsVersion) ? "server" : "--server";
        try {
            var processBuilder = new ProcessBuilder(this.csharpierPath, argument);

            // Setting the working directory is important because older versions of csharpier will accidentally
            // install a recursive file system watch on the entire working directory. If this is the repository root
            // it will include extremely large and unnecessary directories like .git, .idea, bin, obj, node_modules.
            processBuilder.directory(new File(this.csharpierPath).getParentFile());
            processBuilder.environment().put("DOTNET_NOLOGO", "1");
            processBuilder.environment().put("DOTNET_ROOT", this.dotNetProvider.getDotNetRoot());
            var csharpierProcess = processBuilder.start();

            var stdoutThread = new Thread(() -> {
                try (
                    var reader = new BufferedReader(
                        new InputStreamReader(csharpierProcess.getInputStream())
                    )
                ) {
                    var line = reader.readLine();
                    if (line == null) {
                        return;
                    }

                    var portString = line.replace("Started on ", "");
                    this.port = Integer.parseInt(portString);

                    this.logger.debug("Connecting via port " + portString);
                    this.process = csharpierProcess;
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            stdoutThread.start();
            stdoutThread.join();

            csharpierProcess
                .onExit()
                .thenAccept(p -> {
                    try (
                        var errorReader = new BufferedReader(
                            new InputStreamReader(csharpierProcess.getErrorStream())
                        )
                    ) {
                        var error = new StringBuilder();
                        errorReader.lines().forEach(o -> error.append(o + "\n"));
                        this.logger.error("Process failed to start with " + error);
                    } catch (Exception e) {
                        this.logger.error("Process failed to start with " + e);
                    }

                    this.processFailedToStart = true;
                });
        } catch (Exception e) {
            this.logger.warn("Failed to spawn the needed csharpier server.", e);
            this.processFailedToStart = true;
        }
    }

    @Override
    public FormatFileResult formatFile(FormatFileParameter parameter) {
        if (this.processFailedToStart) {
            this.logger.warn("CSharpier process failed to start. Formatting cannot occur.");
            return null;
        }

        var timeWaited = 0;
        while (this.process == null && timeWaited < 15000) {
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {}

            timeWaited += 100;
        }

        if (this.processFailedToStart || this.process == null) {
            this.logger.warn("CSharpier process failed to start. Formatting cannot occur.");
            return null;
        }

        var url = "http://127.0.0.1:" + this.port + "/format";

        try {
            var url1 = URI.create(url).toURL();

            var connection = (HttpURLConnection) url1.openConnection();

            connection.setRequestMethod("POST");

            connection.setRequestProperty("Content-Type", "application/json; utf-8");

            connection.setConnectTimeout(2000);
            connection.setDoOutput(true);
            connection.setDoInput(true);

            var outputStream = connection.getOutputStream();
            var writer = new OutputStreamWriter(outputStream, "UTF-8");
            writer.write(this.gson.toJson(parameter));
            writer.flush();
            writer.close();
            outputStream.close();

            var responseCode = connection.getResponseCode();
            if (responseCode != 200) {
                this.logger.warn(
                        "Csharpier server returned non-200 status code of " + responseCode
                    );
                connection.disconnect();
                return null;
            }

            InputStreamReader reader = new InputStreamReader(connection.getInputStream(), "UTF-8");
            var result = gson.fromJson(reader, FormatFileResult.class);
            reader.close();

            connection.disconnect();

            return result;
        } catch (Exception e) {
            this.logger.warn("Failed posting to the csharpier server.", e);
        }

        return null;
    }

    @Override
    public String getVersion() {
        return this.version;
    }

    @Override
    public boolean getProcessFailedToStart() {
        return this.processFailedToStart;
    }

    @Override
    public String formatFile(String content, String fileName) {
        var parameter = new FormatFileParameter();
        parameter.fileName = fileName;
        parameter.fileContents = content;

        var result = this.formatFile(parameter);
        return result == null ? null : result.formattedFile;
    }

    @Override
    public void dispose() {
        if (this.process != null) {
            this.process.destroy();
        }
    }
}
